packages/xstate-react/src/useInterpret.ts

```diff
@@ -119,6 +125,12 @@ export function useInterpret<
     };
   }, []);
 
+  useIsomorphicLayoutEffect(() => {
+    if (input) {
+      service.input?.(input);
+    }
+  }, [...Object.values(input || {})]);
```

Comment:
Clever!

Reply:
One thing I'm nervous about with this (and the implementation in general) is that passing an un-memoised input to `useInterpret` will not end well:

```ts
input: {
  data: { id: '1', foo: false }
}
```

I _think_ this will end up in an infinite loop, and I'm unsure how to warn for this.

Reply:
Well, it's React world - people should already be ~tired with~ accustomed to memoization. I'm not sure there is much that we can do about this within `@xstate/react`

However, I see a **huge** problem that goes into the performance area of the design here. 

I think this expectation at the end is reasonable:
```js
  it('should not re-render after input changes', () => {
    let renders = 0;
    const model = createModel({}, {}).withInput(
      {} as {
        foo: string;
      }
    );

    const machine = model.createMachine({}, {});

    const App = (props: { foo: string }) => {
      renders++;
      useMachine(machine, {
        input: {
          foo: props.foo
        }
      });
      return null;
    };

    const { rerender } = render(<App foo="1" />);
    rerender(<App foo="2" />);
    
    expect(renders).toBe(2);
  });
```

An input change should not result in a "followup" rerender - and this is now "required" because the `state` changes. The main goal of the feature is to make `input` available **within** the machine, right?

To avoid this rerender problem we'd have to remove `input` from the `state` entirely - because if we keep it there then it should be kept in sync and we can't allow for stale data to be "preserved" there until some next, random, update.

Alternatively, we could bail out from rerendering on input changes and "fake" the `state.input` update in the render phase. This goes into dangerous territory though and opens new challenges related to proper memoization, user expectations and more

Reply:
You're right, this is tricky. It depends if the user ever wants to _read_ `state.input`. I'm not sure what the point of that would be, since by definition you always have the values in scope anyway.

That would mean the input shouldn't be on `meta.state` at all, but `meta.input`.

And the input should live on the `Interpreter`, not the `State`?

Reply:
The only thing why one would like to read `state.input` is convenience - you could pass the whole `state` around instead of 2 separate values if you need access to both of those things. We also should think about how this affects selectors - maybe those should receive "meta" too?

> And the input should live on the Interpreter, not the State?

That's my initial impression as well - maybe some other object should be created within interpreter that would be used for use cases like this. This could live in smth like `actorCtx` that we might need for other stuff too - like the actual `ref` for the "self", maybe also children? 

We probably should allow for this stuff to be serialized for serialization & rehydration purposes - to achieve this goal/use case we don't have to limit ourselves to keeping everything in a single object. Sure, it makes some things easier - but some other things harder. 

This is a very broad discussion though - and I think we could really use a design meeting for this.